File: 1964-find-the-longest-valid-obstacle-course-at-each-position/1964-find-the-longest-valid-obstacle-course-at-each-position.py
class Solution:
    def longestObstacleCourseAtEachPosition(self, obstacles: List[int]) -> List[int]:
        arr, ret = [], []
        for x in obstacles:
            if not arr or arr[-1] <= x: 
                arr.append(x)
                ret.append(len(arr))
            else:
                idx = bisect.bisect(arr, x)
                arr[idx] = x
                ret.append(idx + 1)
        return ret

File: 1964-find-the-longest-valid-obstacle-course-at-each-position/1964-find-the-longest-valid-obstacle-course-at-each-position.cpp
class Solution {
public:
    vector<int> longestObstacleCourseAtEachPosition(vector<int>& obstacles) {
        vector<int> arr;
        int n = obstacles.size();
        vector<int> rets(n);
        
        for(int i = 0; i < n; ++i) {
            if(arr.empty() || arr.back() <= obstacles[i]) {
                arr.push_back(obstacles[i]);
                rets[i] = arr.size();
            } else {
                auto iter = upper_bound(arr.begin(), arr.end(), obstacles[i]);
                
                *iter = obstacles[i];
                rets[i] = iter - arr.begin() + 1;
            }
            
        }
        
        return rets;
    }
};